File: web/src/api/widget.ts
import axios from 'axios';

export type WidgetType = 'note_form' | 'stats' | 'template_form' | 'view' | 'note';

export interface WidgetPosition {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface WidgetData {
  id?: string;
  created_at?: string;
  created_by?: string;
  updated_at?: string;
  updated_by?: string;
  type: WidgetType;
  config?: string;
  position?: string;
}

export interface CreateWidgetData {
  type: WidgetType;
  config?: string;
  position?: string;
}

export interface UpdateWidgetData {
  id: string;
  type?: WidgetType;
  config?: string;
  position?: string;
}

export const getWidgets = async (
  workspaceId: string,
  pageNum: number,
  pageSize: number,
  query?: string,
  type?: WidgetType
) => {
  let url = `/api/v1/workspaces/${workspaceId}/widgets?pageSize=${pageSize}&pageNumber=${pageNum}`;
  if (query) {
    url += `&query=${encodeURIComponent(query)}`;
  }
  if (type) {
    url += `&type=${type}`;
  }
  const response = await axios.get(url, { withCredentials: true });
  return response.data as WidgetData[];
};

export const getWidget = async (workspaceId: string, widgetId: string) => {
  const response = await axios.get(
    `/api/v1/workspaces/${workspaceId}/widgets/${widgetId}`,
    { withCredentials: true }
  );
  return response.data as WidgetData;
};

export const createWidget = async (workspaceId: string, data: CreateWidgetData) => {
  const response = await axios.post(
    `/api/v1/workspaces/${workspaceId}/widgets`,
    data
  );
  return response.data as WidgetData;
};

export const updateWidget = async (workspaceId: string, data: UpdateWidgetData) => {
  const response = await axios.put(
    `/api/v1/workspaces/${workspaceId}/widgets/${data.id}`,
    data
  );
  return response.data as WidgetData;
};

export const deleteWidget = async (workspaceId: string, id: string) => {
  const response = await axios.delete(
    `/api/v1/workspaces/${workspaceId}/widgets/${id}`
  );
  return response.data;
};

File: api/internal/model/widget.go
package model

// WidgetType defines the type of widget
type WidgetType string

const (
	WidgetTypeNoteForm   WidgetType = "note_form"   // Widget for creating/editing notes
	WidgetTypeNoteList   WidgetType = "note_list"   // Widget for displaying notes with conditions
	WidgetTypeNote       WidgetType = "note"        // Widget for displaying a single note's complete content
	WidgetTypeCountdown  WidgetType = "countdown"   // Widget for countdown timer
	WidgetTypeFileUpload WidgetType = "file_upload" // Widget for uploading files
	WidgetTypeCarousel   WidgetType = "carousel"    // Widget for image carousel
	WidgetTypeRSS        WidgetType = "rss"         // Widget for displaying RSS feeds
	WidgetTypeMusic      WidgetType = "music"       // Widget for playing audio files
	WidgetTypeVideo      WidgetType = "video"       // Widget for playing video files
	WidgetTypeIframe     WidgetType = "iframe"      // Widget for embedding external web pages
	WidgetTypeFolder     WidgetType = "folder"      // Widget for organizing widgets in a folder
	WidgetTypeLink       WidgetType = "link"        // Widget for displaying a clickable link
	WidgetTypeMap        WidgetType = "map"         // Widget for displaying a map view
	WidgetTypeCalendar   WidgetType = "calendar"    // Widget for displaying a calendar view
)

type WidgetFilter struct {
	WorkspaceID      string
	WidgetIDs        string
	Type             string
	ParentID         string // Filter by parent widget ID (empty string = root widgets, "*" = all widgets)
	PageSize         int
	PageNumber       int
	Query            string
}

type Widget struct {
	WorkspaceID string `json:"workspace_id"`
	ID          string `json:"id"`
	Type        string `json:"type"`     // Widget type (note_form, stats, template_form, view, note_list, note, folder)
	Config      string `json:"config"`   // JSON configuration for the widget
	Position    string `json:"position"` // JSON position data (x, y, width, height)
	ParentID    string `json:"parent_id"` // Parent widget ID for hierarchical structure (null for root widgets)
	CreatedAt   string `json:"created_at"`
	CreatedBy   string `json:"created_by"`
	UpdatedAt   string `json:"updated_at"`
	UpdatedBy   string `json:"updated_by"`
}